### ISSUE TITLE
Mypy cannot infer types of initialized variables

### ISSUE THREAD

**Issue by KaiqueBressi** [python/mypy#8493]
I have the following code:

```python
class Service:
    def __init__():
        self.another_service = Service()
     
     def my_method() -> None:
         return self.another_serive.another_method_call()
```

Mypy should gives an error of incompatible types because `another_method_call` is not returnig `None` but it is not type checking because it cannot identify what is the type of `another_service`.

The only way to makes it gives an error is initializing `another_service` using:

```python
self.another_service: Service = Service() 
```

I think mypy has enough static information to guess what is the type of `another_service` wihout needing to someone explicit type it.

**Comment by JelleZijlstra** [comment]
Does it work if you add a type annotation to `__init__`? Mypy generally doesn't typecheck unannotated methods.

**Comment by ilevkivskyi** [comment]
Adding a return annotation to `__init__()` indeed fixes the issue. I however still think it is a bug. We had a fix a while ago that introduced type inference for `self` attributes in non-annotated methods. It looks like it doesn't work with deferred methods.

cc @msullivan

**Comment by KaiqueBressi** [comment]
> Adding a return annotation to `__init__()` indeed fixes the issue. I however still think it is a bug. We had a fix a while ago that introduced type inference for `self` attributes in non-annotated methods. It looks like it doesn't work with deferred methods.
> 
> cc @msullivan

Yeah, I was a little frustrated because forgetting to add a type annotation to `__init__` would cause type annotated methods that depends on `__init__` to ignore any typing error, which is not good but using the command line ` --check-untyped-defs` is a good temporary solution for now. 

Thanks for the responses.
